packages/pubsub/src/batch.js

```diff
@@ -0,0 +1,220 @@
+ * @param {number} options.maxBytes - Max size of data to be sent per request.
+ * @param {number} options.maxMessages - Max number of messages to be sent per
+ *     request.
+ * @param {number} options.maxMilliseconds - Max time to wait before sending
+ *     data.
+ * @param {number} options.maxRequests - Max number of concurrent requests
+ *     allowed.
+ * @param {function(messages, callback)} options.sendHandler - Function to be
+ *     called with message batch.
+ *
+ * @example
+ * var batch = new Batch({
+ *   maxBytes: Math.pow(1024, 2) * 5,
+ *   maxMessages: 1000,
+ *   maxMilliseconds: 1000,
+ *   maxRequests: 5,
+ *   send: function(messages, done) {
+ *     var reqOpts = {
+ *       topic: self.topic.name,
+ *       messages = messages.map(function(message) {
+ *         return {
+ *           data: message.data,
+ *           attributes: message.attrs
+ *         };
+ *       })
+ *     };
+ *
+ *     return self.api.Publisher.publish(reqOpts, function(err, resp) {
+ *       if (err) {
+ *         done(err);
+ *         return;
+ *       }
+ *
+ *       done(null, {
+ *         responses: resp.messageIds
+ *       })
+ *     });
+ *   }
+ * });
+ */
+function Batch(options) {
+  this.maxBytes = options.maxBytes || Math.pow(1024, 2) * 5;
+  this.maxMessages = options.maxMessages || 1000;
+  this.maxMilliseconds = options.maxMilliseconds || 1000;
+  this.maxRequests = options.maxRequests || 5;
+  this.sendHandler = options.send;
+
+  this.inventory = {
+    queued: new Set(),
+    inFlight: new Set()
+  };
+
+  Object.defineProperty(this.inventory, 'queueBytes', {
+    get: function() {
+      var size = 0;
+
+      this.queued.forEach(function(message) {
+        size += message.size;
+      });
+
+      return size;
+    }
+  });
+
+  this.intervalHandle_ = null;
+  this.activeRequests_ = 0;
+}
+
+/**
+ * Adds message to the queue.
+ *
+ * @param {*} data - The message data.
+ * @param {function} callback - Callback to be ran once message is sent.
+ *
+ * @example
+ * batch.add(1234, function(err, resp) {});
+ */
+Batch.prototype.add = function(data, callback) {
+  this.inventory.queued.add({
+    data: data,
+    callback: callback
+  });
+
+  var reachedMaxMessages = this.inventory.queued.size >= this.maxMessages;
+  var reachedMaxBytes = this.inventory.queueBytes >= this.maxBytes;
+  var reachedMaxRequests = this.activeRequests_ >= this.maxRequests;
+
+  if ((reachedMaxMessages || reachedMaxBytes) && !reachedMaxRequests) {
+    this.send(this.getNextMessageBatch());
+  } else if (!this.intervalHandle_) {
+    this.beginSending();
+  }
+};
+
+/**
+ * Starts the send interval.
+ *
+ * @example
+ * batch.beginSending();
+ */
+Batch.prototype.beginSending = function() {
+  if (this.intervalHandle_) {
+    return;
+  }
+
+  var self = this;
+
+  this.intervalHandle_ = setInterval(function() {
+    self.send(self.getNextMessageBatch());
+
+    if (self.inventory.queued.size === 0) {
```

Comment:
I'm surprised getNextMessageBatch doesn't always flush the queue. Why wouldn't it?